drud/ddev
# Use simple container name as hostname inside docker network (use "db" instead of "ddev-<project>-db")

## The Problem/Issue/Bug:

* In recent versions of docker-compose there have been problems with the `links` stanza, where it does not properly bind "db" for example, to the project's "db" service. These have been solved in the docker-compose that is currently bundled by ddev.
* But for years (since the beginning of ddev) containers included without a "links" (like solr) have resulted in ambiguity and unpredictable behavior between projects. One ddev project might accidentally connect to another project's solr service, for example.

Unambiguity in name resolution was recently only possible in `ddev`when using `ddev-<projectname>-<service>` which requires a level of environment-awareness in apps that should usually not be necessary in a container environment.

## How this PR Solves The Problem:

This PR sets up networking where the service name of a project's container is perfectly adequate to address it. For example, `ping db` in the web container will always get the related project's db container. `ping solr` will always get the related solr container, not one from another service.

This makes ddev use  standard docker-compose networking as described in https://docs.docker.com/compose/networking/.

The "ddev" network is now only used to communicate across projects and with the proxy.

Contrib-service-recipes that only need per-project networking (via its service name or `ddev-<projectname>-<service>:<port>`) will not need any changes.
Those that use HTTP_EXPOSE and HTTPS_EXPOSE to be reachable externally via `http[s]://<project>:<port>` (via ddev-router) or cross-project need to add the "ddev" network:
```
networks: [default, ddev]
```
This can be given as general advise, too because it does not hurt.

All ddev-contrib service examples should be updated.

## Manual Testing Instructions:

To test this: 
* Use [gitpod](https://gitpod.io/#https://github.com/drud/ddev/pull/3403)
* Download the artifacts in the latest [build](https://github.com/drud/ddev/actions/workflows/pr-build.yml?query=branch%3A20211128__docker-compose-networking++) and [put ddev in your path](https://github.com/drud/ddev/blob/master/docs/developers/building-contributing.md#testing-a-pr)

Create multiple ddev projects (`ddev config` with all defaults).
Run `ddev exec ping db` from them.

You should see answers only from the project-local "db" container.
With docker-compose v2.1.1 you would see answers from other "db" containers.

In order to test cross-project connectivity, run 
* `ddev exec ping ddev-<other-projectname>-db`.

To test this properly, use docker-compose v2.1.1, which had the bug about not respecting `links`. `ddev config global --docker-compose-version=v2.1.1`


## Automated Testing Overview:

- [x] Added TestNetworkAmbiguity

## Related Issue Link(s):

* #3369
* docker/compose#8935

## Remarks/tbd:
* This change tries to minimize code changes. For improved self-documentation the network should rather be called "ddev_proxy" or "ddev_frontend".
- [x] Removed ddev-ssh-agent networks completely
- [x] Usage of docker-compose service "links" should not be necessary any more.
- [ ] ddev-contrib services (especially containing HTTP_EXPOSE) needs to be updated with the new networking on release
- [ ] Release notes need to explain the change and what it means to 3rd party projects

<a href="https://gitpod.io/#https://github.com/drud/ddev/pull/3403"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
